src/Conversion/ONNXToKrnl/Math/Elementwise.cpp

```diff
@@ -1050,11 +1051,27 @@ template <>
 Value emitScalarOpFor<ONNXEqualOp>(ConversionPatternRewriter &rewriter,
     Location loc, Operation *op, Type elementType,
     ArrayRef<Value> scalarOperands) {
+
   CheckIfCustomScalarOpIsSupported<ONNXEqualOp>(elementType);
+  Value results;
   Value lhs = scalarOperands[0];
   Value rhs = scalarOperands[1];
-  MultiDialectBuilder<MathBuilder> create(rewriter, loc);
-  return create.math.eq(lhs, rhs);
+  MultiDialectBuilder<KrnlBuilder, MathBuilder> create(rewriter, loc);
+  Type inputElemType = getElementType(lhs.getType());
+
+  // If the two input values are a string then we want to use the krnlstrncmp.
+  // However, if the input values are a float or an int we can simply use the
+  // equal function.
+  if (inputElemType.isa<krnl::StringType>()) {
+    Value strlenRes = create.krnl.strlen(lhs);
+    Value strncmpRes = create.krnl.strncmp(lhs, rhs, strlenRes);
```

Comment:
Just to confirm one thing here: when two strings are equal, `strncmp` returns `zero` (https://en.cppreference.com/w/c/string/byte/strncmp), it means ONNXEqual returns `zero`. But I guess MLIR interprets `zero` as `false`, doesn't it?

Reply:
@tungld Is this update better?

Reply:
will this say the strings are equal when lhs is a prefix of rhs? I wonder if we should call `strcmp` instead of `strncmp`, or call `strncmp` with `strlenRes+1`

Reply:
@sorenlassen in `onnx-mlir/src/Dialect/Krnl/DialectBuilder.cpp` the only method we have defined is `strncmp` when we are doing the lowering to krnl. So that is why I used it. According to my search:

`The basic difference between these two are : strcmp compares both the strings till null-character of either string comes whereas strncmp compares at most num characters of both strings.`

So I do not see a huge advantage in terms of performance but if you see the benefit of making the change...let me know. Thanks!

Reply:
since you only only compare at most strlenRes then I think you don't distinguish between the cases where lhs is only a prefix of rhs or they are equal

a quick fix could be to pass strlenRes+1 instead of strlenRes as the 3rd argument to strncmp